### ISSUE TITLE
provide selection_line_width and nonselection_line_width

### ISSUE THREAD

**Issue by solstag** [bokeh/bokeh#8893]
**Is your feature request related to a problem? Please describe.**
I would like to be able to set a line width for selections. In my case these are "rect", so if implemented this would allow a nice effect of a "frame" around the selection.

**Describe the solution you'd like**
I woud like for `selection_line_width` to be usable just like i.e. `selection_line_color`.
I would be willing to try to implement this if I get some guidance on how to proceed.

**Additional context**
In my particular application I have adjacent rectangles which are densely packed in some areas and spread out in other areas, so setting line_width to a larger value and playing with selection_line_alpha is not an option.

**Comment by bryevdv** [comment]
This is a duplicate of #2367

**Comment by solstag** [comment]
Ni! Thanks @bryevdv, I'm a bit confused as to why this is a duplicate. In that issue it is said that:

> Current behavior is as designed and documented, i.e. only "visual" (line and fill) props are considered for selection/nonselection/hover policies.

In what sense isn't line_width "visual", then? Cheers

**Comment by bryevdv** [comment]
@solstag I misunderstood what you were asking about. The relevant code path is here:

https://github.com/bokeh/bokeh/blob/c3a9d4d5ab5662ea34813fd72de50e1bdaae714d/bokeh/plotting/helpers.py#L839-L843

The `_pop_colors_and_alpha` could be renamed and generalized to `_pop_visuals` and inspect the passed-in glyph class to handle any visuals present on the glyph. A new `_pop_visuals` would still need to handle the special cases for "color" being an alias for setting *both* "line_color" and "fill_color".

**Comment by solstag** [comment]
Ni!

Ok, I have worked it out. However, before I publish the patch and make a pull request, there's a compatibility issue I want to discuss.

Currently, if you set a non prefixed visual property handled by `_pop_colors_and_alpha` for a glyph, i.e. `line_color`, it does not affect any prefixed glyphs since the key is popped from the dictionary.

However, all other visual properties, i.e. `line_width`, are currently copied from the non prefixed glyph because they are not popped and passed on inside `kws` to `_make_glyph`.

1) I believe the more desirable behavior is for the prefixed glyph to inherit from the non prefixed glyph everything that has not been explicitly declared. However this would break compatibility for properties currently treated by `_pop_colors_and_alpha`.

2) Another option is to not inherit anything from the prefixed glyph. This would also break compatibility, but for the properties not currently treated, which would no longer be inherited from the non prefixed properties.

3) The third solution would be to force compatibility by selectively treating the currently handled properties. I personally think this is undesirable.

I have the two first options working, and the third is of course feasible, but ugly.

Any ideas on how to handle this ? If it doesn't seem like a big issue, I would go for (1). I think giving prefixes the semantics of "this should be different from the basic (unprefixed) condition" makes more sense in bokeh usage.

(Like, if I style a glyph with several visual properties and then set a single `selection_` property, I expect the selected glyph to look just like my regular styled glyph, except for that one thing, not have it look like the default glyph.)

However there may be deeper compatibility issues I am not aware of. So what do you say?!

**Comment by bryevdv** [comment]
@solstag I am going to open long running a branch for 2.0 development very soon. It's not our usual process style to do PRs against anything other than `master` but for this case, since we are dropping Python 2 support and making other incompatible changes, it seems necessary if we are to be able to track PRs for individual work.

So, my thoughts would be this: I would be happy to accept a PR from you for approach 1) against the future 2.0 branch. Does that sound good? FWIW I would like to have 2.0 within the next 3 months.

**Comment by solstag** [comment]
Hi @bryevdv, I'm in a busy period so that looks like a good time frame for me. I'm happy to contribute to the migration guide once the PR is merged. Happy to hear Python2 is going into (well deserved) retirement. ;-)

**Comment by bryevdv** [comment]
@solstag fantastic, FYI the branch to make a PR against when you are ready is `landing-2.0`

**Comment by solstag** [comment]
Great. For now, I've published a branch based on landing-2.0 with my current changes rebased:

https://github.com/solstag/bokeh/tree/style_visuals

**Comment by bryevdv** [comment]
@solstag can you go ahead and submit a PR for your branch? I want to make sure this does not get forgotten before 2.0

**Comment by solstag** [comment]
Thanks for the reminder @bryevdv , will do it next week.

**Comment by bryevdv** [comment]
@solstag did you still want to submit a PR for this against the landing-2.0 branch (please make sure to pull everything up to date)

**Comment by solstag** [comment]
Yes @bryevdv , sorry I've been absent. Will push soon.

**Comment by solstag** [comment]
Two questions...

- I'm assuming we want the "trait" aliases (ex. 'color') to affect all visuals. The only case I'm unsure of is with 'color' affecting 'text_color'. I don't know in what glyphs this could interact with other colors and possibly be undesirable. But if there's no obvious conflict, I would keep the alias rule simple.

- Must I also convert `_pop_colors_and_alpha` in `bokehjs/src/lib/api/plotting.ts` so both functions have the same logic? I get that it is TypeScript, right? I'm new to it, but I might be able to figure it out given a little more time.

(I have pushed my latest changes to [my branch](https://github.com/solstag/bokeh/commits/style_visuals), they're pretty much ready to go if not for the two issues above).

**Comment by bryevdv** [comment]
for glyph methods, `text_color` only pertains to the `Text` glyph (there are other things with text color properties, but they are not glyphs, so not covered by this work) I think keeping it simple and having `color` set every `*_color` property is exactly right.

Re: `plotting.ts` if you are up for it, yes ideally a PR will update that logic to match. We could accept a PR without that update, as long as there was a new issue made to update that file, later.

Can you go ahead and submit a PR based on the current branch? If any changes are necessary we can discuss them more easily in the PR
